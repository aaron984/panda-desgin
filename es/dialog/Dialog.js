import _extends from "@babel/runtime/helpers/extends";
import _objectSpread from "@babel/runtime/helpers/objectSpread";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from 'react';
import Animate from '../AnimationGroup';
import LazyRenderBox from './LazyRenderBox';

var Dialog =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Dialog, _React$Component);

  function Dialog() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Dialog);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Dialog)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "getDialogElement", function () {
      var props = _this.props;
      var closable = props.closable;
      var prefixCls = props.prefixCls;
      var footer;

      if (props.footer) {
        footer = React.createElement("div", {
          className: "".concat(prefixCls, "-footer"),
          ref: function ref(el) {
            return _this.footerRef = el;
          }
        }, props.footer);
      }

      var header;

      if (props.title) {
        header = React.createElement("div", {
          className: "".concat(prefixCls, "-header")
        }, React.createElement("div", {
          className: "".concat(prefixCls, "-title")
        }, props.title));
      }

      var closer;

      if (closable) {
        closer = React.createElement("button", {
          onClick: _this.close,
          "aria-label": "Close",
          className: "".concat(prefixCls, "-close")
        }, React.createElement("span", {
          className: "".concat(prefixCls, "-close-x")
        }));
      }

      var transitionName = _this.getTransitionName();

      var dialogElement = React.createElement(LazyRenderBox, {
        key: "dialog-element",
        role: "document",
        ref: function ref(el) {
          return _this.dialogRef = el;
        },
        style: props.style || {},
        className: "".concat(prefixCls, " ").concat(props.className || ''),
        visible: props.visible
      }, React.createElement("div", {
        className: "".concat(prefixCls, "-content")
      }, closer, header, React.createElement("div", {
        className: "".concat(prefixCls, "-body"),
        style: props.bodyStyle,
        ref: function ref(el) {
          return _this.bodyRef = el;
        }
      }, props.children), footer));
      return React.createElement(Animate, {
        key: "dialog",
        showProp: "visible",
        onAppear: _this.onAnimateAppear,
        onLeave: _this.onAnimateLeave,
        transitionName: transitionName,
        component: "",
        transitionAppear: true
      }, dialogElement);
    });

    _defineProperty(_assertThisInitialized(_this), "onAnimateAppear", function () {
      document.body.style.overflow = 'hidden';
    });

    _defineProperty(_assertThisInitialized(_this), "onAnimateLeave", function () {
      document.body.style.overflow = '';
      /* istanbul ignore else */

      if (_this.wrapRef) {
        _this.wrapRef.style.display = 'none';
      }

      _this.props.onAnimateLeave();

      _this.props.afterClose();
    });

    _defineProperty(_assertThisInitialized(_this), "close", function (e) {
      _this.props.onClose(e);
    });

    _defineProperty(_assertThisInitialized(_this), "onMaskClick", function (e) {
      if (e.target === e.currentTarget) {
        _this.close(e);
      }
    });

    return _this;
  }

  _createClass(Dialog, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      // fix: react@16 no dismissing animation
      document.body.style.overflow = '';
      /* istanbul ignore else */

      if (this.wrapRef) {
        this.wrapRef.style.display = 'none';
      }
    }
  }, {
    key: "getZIndexStyle",
    value: function getZIndexStyle() {
      var style = {};
      var props = this.props;

      if (props.zIndex !== undefined) {
        style.zIndex = props.zIndex;
      }

      return style;
    }
  }, {
    key: "getWrapStyle",
    value: function getWrapStyle() {
      var wrapStyle = this.props.wrapStyle || {};
      return _objectSpread({}, this.getZIndexStyle(), wrapStyle);
    }
  }, {
    key: "getMaskStyle",
    value: function getMaskStyle() {
      var maskStyle = this.props.maskStyle || {};
      return _objectSpread({}, this.getZIndexStyle(), maskStyle);
    }
  }, {
    key: "getMaskTransitionName",
    value: function getMaskTransitionName() {
      var props = this.props;
      var transitionName = props.maskTransitionName;
      var animation = props.maskAnimation;

      if (!transitionName && animation) {
        transitionName = "".concat(props.prefixCls, "-").concat(animation);
      }

      return transitionName;
    }
  }, {
    key: "getTransitionName",
    value: function getTransitionName() {
      var props = this.props;
      var transitionName = props.transitionName;
      var animation = props.animation;

      if (!transitionName && animation) {
        transitionName = "".concat(props.prefixCls, "-").concat(animation);
      }

      return transitionName;
    }
  }, {
    key: "getMaskElement",
    value: function getMaskElement() {
      var props = this.props;
      var maskElement;

      if (props.mask) {
        var maskTransition = this.getMaskTransitionName();
        maskElement = React.createElement(LazyRenderBox, _extends({
          style: this.getMaskStyle(),
          key: "mask-element",
          className: "".concat(props.prefixCls, "-mask"),
          hiddenClassName: "".concat(props.prefixCls, "-mask-hidden"),
          visible: props.visible
        }, props.maskProps));

        if (maskTransition) {
          maskElement = React.createElement(Animate, {
            key: "mask",
            showProp: "visible",
            transitionAppear: true,
            component: "",
            transitionName: maskTransition
          }, maskElement);
        }
      }

      return maskElement;
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var props = this.props;
      var prefixCls = props.prefixCls,
          maskClosable = props.maskClosable;
      var style = this.getWrapStyle();

      if (props.visible) {
        style.display = null;
      }

      return React.createElement("div", null, this.getMaskElement(), React.createElement("div", _extends({
        className: "".concat(prefixCls, "-wrap ").concat(props.wrapClassName || ''),
        ref: function ref(el) {
          return _this2.wrapRef = el;
        },
        onClick: maskClosable ? this.onMaskClick : undefined,
        role: "dialog",
        "aria-labelledby": props.title,
        style: style
      }, props.wrapProps), this.getDialogElement()));
    }
  }]);

  return Dialog;
}(React.Component);

_defineProperty(Dialog, "defaultProps", {
  className: '',
  mask: true,
  visible: false,
  closable: true,
  maskClosable: true,
  prefixCls: 'panda-dialog',
  onClose: function onClose() {},
  onAnimateLeave: function onAnimateLeave() {},
  afterClose: function afterClose() {}
});

export { Dialog as default };