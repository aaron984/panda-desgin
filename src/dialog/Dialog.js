import React from 'react';
import PropTypes from 'prop-types';
import Animate from '../AnimationGroup';
import LazyRenderBox from './LazyRenderBox';

const propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.element,
  ]).isRequired,
  footer: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.element,
  ]),
  title: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.element,
  ]),
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  mask: PropTypes.bool,
  visible: PropTypes.bool,
  closable: PropTypes.bool,
  maskClosable: PropTypes.bool,
  style: PropTypes.objectOf(PropTypes.string),
  bodyStyle: PropTypes.objectOf(PropTypes.string),
  onClose: PropTypes.func,
  onAnimateLeave: PropTypes.func,
  afterClose: PropTypes.func,
};

export default class Dialog extends React.Component {
  static propTypes = propTypes

  static defaultProps = {
    mask: true,
    visible: false,
    closable: true,
    maskClosable: true,
    prefixCls: 'panda-dialog',
    onClose() { },
    onAnimateLeave() { },
    afterClose() {},
  };

  componentWillUnmount() {
    // fix: react@16 no dismissing animation
    document.body.style.overflow = '';
    /* istanbul ignore else */
    if (this.wrapRef) {
      this.wrapRef.style.display = 'none';
    }
  }

  getZIndexStyle() {
    const style = {};
    const props = this.props;
    if (props.zIndex !== undefined) {
      style.zIndex = props.zIndex;
    }
    return style;
  }

  getWrapStyle() {
    const wrapStyle = this.props.wrapStyle || {};
    return { ...this.getZIndexStyle(), ...wrapStyle };
  }

  getMaskStyle() {
    const maskStyle = this.props.maskStyle || {};
    return { ...this.getZIndexStyle(), ...maskStyle };
  }

  getMaskTransitionName() {
    const props = this.props;
    let transitionName = props.maskTransitionName;
    const animation = props.maskAnimation;
    if (!transitionName && animation) {
      transitionName = `${props.prefixCls}-${animation}`;
    }
    return transitionName;
  }

  getTransitionName() {
    const props = this.props;
    let transitionName = props.transitionName;
    const animation = props.animation;
    if (!transitionName && animation) {
      transitionName = `${props.prefixCls}-${animation}`;
    }
    return transitionName;
  }

  getMaskElement() {
    const props = this.props;
    let maskElement;
    if (props.mask) {
      const maskTransition = this.getMaskTransitionName();
      maskElement = (
        <LazyRenderBox
          style={this.getMaskStyle()}
          key="mask-element"
          className={`${props.prefixCls}-mask`}
          hiddenClassName={`${props.prefixCls}-mask-hidden`}
          visible={props.visible}
          {...props.maskProps}
        />
      );
      if (maskTransition) {
        maskElement = (
          <Animate
            key="mask"
            showProp="visible"
            transitionAppear
            component=""
            transitionName={maskTransition}
          >
            {maskElement}
          </Animate>
        );
      }
    }
    return maskElement;
  }

  getDialogElement = () => {
    const props = this.props;
    const closable = props.closable;
    const prefixCls = props.prefixCls;

    let footer;
    if (props.footer) {
      footer = (
        <div className={`${prefixCls}-footer`} ref={el => this.footerRef = el}>
          {props.footer}
        </div>
      );
    }

    let header;
    if (props.title) {
      header = (
        <div className={`${prefixCls}-header`}>
          <div className={`${prefixCls}-title`}>
            {props.title}
          </div>
        </div>
      );
    }

    let closer;
    if (closable) {
      closer = (
        <button
          onClick={this.close}
          aria-label="Close"
          className={`${prefixCls}-close`}
        >
          <span className={`${prefixCls}-close-x`} />
        </button>
      );
    }

    const transitionName = this.getTransitionName();
    const dialogElement = (
      <LazyRenderBox
        key="dialog-element"
        role="document"
        ref={el => this.dialogRef = el}
        style={props.style || {}}
        className={`${prefixCls} ${props.className || ''}`}
        visible={props.visible}
      >
        <div className={`${prefixCls}-content`}>
          {closer}
          {header}
          <div
            className={`${prefixCls}-body`}
            style={props.bodyStyle}
            ref={el => this.bodyRef = el}
          >
            {props.children}
          </div>
          {footer}
        </div>
      </LazyRenderBox>
    );
    return (
      <Animate
        key="dialog"
        showProp="visible"
        onAppear={this.onAnimateAppear}
        onLeave={this.onAnimateLeave}
        transitionName={transitionName}
        component=""
        transitionAppear
      >
        {dialogElement}
      </Animate>
    );
  }

  onAnimateAppear = () => {
    document.body.style.overflow = 'hidden';
  }

  onAnimateLeave = () => {
    document.body.style.overflow = '';
    /* istanbul ignore else */
    if (this.wrapRef) {
      this.wrapRef.style.display = 'none';
    }

    this.props.onAnimateLeave();
    this.props.afterClose();
  }

  close = (e) => {
    this.props.onClose(e);
  }

  onMaskClick = (e) => {
    if (e.target === e.currentTarget) {
      this.close(e);
    }
  }

  render() {
    const { props } = this;
    const { prefixCls, maskClosable } = props;
    const style = this.getWrapStyle();
    if (props.visible) {
      style.display = null;
    }
    return (
      <div>
        {this.getMaskElement()}
        <div
          className={`${prefixCls}-wrap ${props.wrapClassName || ''}`}
          ref={el => this.wrapRef = el}
          onClick={maskClosable ? this.onMaskClick : undefined}
          role="dialog"
          aria-labelledby={props.title}
          style={style}
          {...props.wrapProps}
        >
          {this.getDialogElement()}
        </div>
      </div>
    );
  }
}
