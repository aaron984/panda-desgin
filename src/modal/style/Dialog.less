.@{modalPrefixClass} {
  position: relative;

  &:not(.@{modalPrefixClass}-transparent):not(.@{modalPrefixClass}-popup) {
    width: 100%;
    height: 100%;
  }

  &-mask {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: @modal-zindex;
    height: 100%;
    background-color: @fill-mask;

    &-hidden {
      display: none;
    }
  }

  &-wrap {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: @modal-zindex;
    display: flex;
    align-items: center;
    justify-content: center;
    height: 100%;
    overflow: auto;
    outline: 0;
    // fixed a layer issue with animated Tabs in the x5 kernel browser
    transform: translateZ(1px);
    -webkit-overflow-scrolling: touch;

    &-popup {
      display: block;
    }
  }

  &-transparent {
    width: 270 * @hd;

    .@{modalPrefixClass}-content {
      border-radius: @radius-lg;
    }
  }

  &-popup {
    position: fixed;
    left: 0;
    width: 100%;

    .@{modalPrefixClass}-content {
      border-radius: 0;
    }

    &-slide-down {
      top: 0;
    }

    &-slide-up {
      bottom: 0;
    }
  }

  &-title {
    margin: 0;
    font-size: @modal-font-size-heading;
    line-height: @line-height-base;
    text-align: center;
    color: @color-text-base;
  }

  &-header {
    padding: @v-spacing-sm @h-spacing-lg @v-spacing-lg;
  }

  &-content {
    position: relative;
    height: 100%;
    overflow: hidden;
    text-align: center;
    background-color: @fill-base;
    border: 0;
    background-clip: padding-box;
  }

  &-close {
    position: absolute;
    right: @h-spacing-lg;
    z-index: 999;
    width: @icon-size-sm;
    height: @icon-size-sm;
    padding: 0;
    background-color: transparent;
    border: 0;
    outline: none;

    &-x {
      display: inline-block;
      width: 15 * @hd;
      height: 15 * @hd;
      background-repeat: no-repeat;
      background-size: cover;
      .encoded-svg-background('modal_delete');
    }
  }

  &-body {
    height: 100%;
    overflow: auto;
    line-height: @line-height-paragraph;
  }
}
