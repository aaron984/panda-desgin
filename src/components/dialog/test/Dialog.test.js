import React from 'react';
import { render, mount } from 'enzyme';
import Dialog from '../Dialog';
import { wrap } from 'module';

describe('Dialog', () => {
  it('Dialog renders correctly', () => {
    const props = {
      visible: false,
      maskAnimation: 'test-maskAnimation',
      title: 'title',
      zIndex: 10,
      animation: 'test-animation',
      closable: false,
    };

    const wrapper = render(<Dialog {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Dialog mask is false renders correctly', () => {
    const props = {
      mask: false,
    };

    const wrapper = render(<Dialog {...props} />);
    expect(wrapper.find('.panda-dialog-mask')).toHaveLength(0);
  });

  it('Dialog maskClosable is true renders correctly', () => {
    const onClose = jest.fn();
    const wrapper = mount(<Dialog onClose={onClose} visible maskClosable />);

    wrapper.find('.panda-dialog-wrap').simulate('click');
    expect(onClose).toHaveBeenCalled();
  });

  it('Dialog onMaskClick should work fine', () => {
    const onClose = jest.fn();
    const wrapper = mount(<Dialog onClose={onClose} visible maskClosable footer="footer" />);

    wrapper.find('.panda-dialog-footer').simulate('click');
    expect(onClose).not.toHaveBeenCalled();
  });

  it('Dialog maskClosable is false renders correctly', () => {
    const onClose = jest.fn();
    const wrapper = mount(<Dialog onClose={onClose} visible maskClosable={false} />);

    wrapper.find('.panda-dialog-wrap').simulate('click');
    expect(onClose).not.toHaveBeenCalled();
  });

  it('Dialog onAnimateLeave function should work fine', () => {
    const onAnimateLeave = jest.fn();
    const afterClose = jest.fn();
    const wrapper = mount(<Dialog onAnimateLeave={onAnimateLeave} afterClose={afterClose} visible />);
    const instance = wrapper.instance();

    instance.onAnimateLeave();
    expect(onAnimateLeave).toHaveBeenCalled();
    expect(afterClose).toHaveBeenCalled();
  });
});
