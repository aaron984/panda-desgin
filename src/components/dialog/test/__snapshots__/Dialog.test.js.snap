// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog Dialog renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "panda-dialog-mask panda-dialog-mask-hidden",
          "style": "z-index:10",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "aria-labelledby": "title",
            "class": "panda-dialog-wrap ",
            "role": "dialog",
            "style": "z-index:10",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "panda-dialog ",
                "role": "document",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "panda-dialog-content",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "class": "panda-dialog-header",
                      },
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "class": "panda-dialog-title",
                          },
                          "children": Array [
                            Object {
                              "data": "title",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "text",
                            },
                          ],
                          "name": "div",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "class": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "class": undefined,
                          },
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": Object {
                        "attribs": Object {
                          "class": "panda-dialog-body",
                        },
                        "children": Array [],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": [Circular],
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "class": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "class": undefined,
                        },
                      },
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                    Object {
                      "attribs": Object {
                        "class": "panda-dialog-body",
                      },
                      "children": Array [],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": Object {
                        "attribs": Object {
                          "class": "panda-dialog-header",
                        },
                        "children": Array [
                          Object {
                            "attribs": Object {
                              "class": "panda-dialog-title",
                            },
                            "children": Array [
                              Object {
                                "data": "title",
                                "next": null,
                                "parent": [Circular],
                                "prev": null,
                                "type": "text",
                              },
                            ],
                            "name": "div",
                            "namespace": "http://www.w3.org/1999/xhtml",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": Object {
                              "class": undefined,
                            },
                            "x-attribsPrefix": Object {
                              "class": undefined,
                            },
                          },
                        ],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": [Circular],
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "class": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "class": undefined,
                        },
                      },
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "role": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "role": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "aria-labelledby": undefined,
            "class": undefined,
            "role": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "aria-labelledby": undefined,
            "class": undefined,
            "role": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
      Object {
        "attribs": Object {
          "aria-labelledby": "title",
          "class": "panda-dialog-wrap ",
          "role": "dialog",
          "style": "z-index:10",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "panda-dialog ",
              "role": "document",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "panda-dialog-content",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "panda-dialog-header",
                    },
                    "children": Array [
                      Object {
                        "attribs": Object {
                          "class": "panda-dialog-title",
                        },
                        "children": Array [
                          Object {
                            "data": "title",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "text",
                          },
                        ],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "class": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "class": undefined,
                        },
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Object {
                      "attribs": Object {
                        "class": "panda-dialog-body",
                      },
                      "children": Array [],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  Object {
                    "attribs": Object {
                      "class": "panda-dialog-body",
                    },
                    "children": Array [],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": Object {
                      "attribs": Object {
                        "class": "panda-dialog-header",
                      },
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "class": "panda-dialog-title",
                          },
                          "children": Array [
                            Object {
                              "data": "title",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "text",
                            },
                          ],
                          "name": "div",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "class": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "class": undefined,
                          },
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "role": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "role": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "panda-dialog-mask panda-dialog-mask-hidden",
            "style": "z-index:10",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-labelledby": undefined,
          "class": undefined,
          "role": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-labelledby": undefined,
          "class": undefined,
          "role": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
